linker.go

```diff
@@ -0,0 +1,386 @@
+
+			path := uuid.New().String()
+			o, err := l.CreateLink(path, firstTarget)
+
+			defer func() {
+				err = store.Delete(path)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			Convey("The first returned error should be nil", func() {
+				So(err, ShouldBeNil)
+			})
+
+			secondSize := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
+			secondContent, _ := ioutil.ReadAll(io.LimitReader(randbytes.NewRand(), secondSize))
+			secondTarget := uuid.New().String()
+
+			_, err = store.Write(secondTarget, bytes.NewReader(secondContent), secondSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(secondTarget)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			o, err = l.CreateLink(path, secondTarget)
+
+			Convey("The second returned error should also be nil", func() {
+				So(err, ShouldBeNil)
+			})
+
+			Convey("The object mode should be link", func() {
+				// Link object's mode must be link.
+				So(o.Mode.IsLink(), ShouldBeTrue)
+			})
+
+			Convey("The linkTarget of the object must be the same as the secondTarget", func() {
+				// The linkTarget must be the same as the secondTarget.
+				linkTarget, ok := o.GetLinkTarget()
+
+				So(ok, ShouldBeTrue)
+				So(linkTarget, ShouldEqual, secondTarget)
+			})
+
+			Convey("Read should get path object data without error", func() {
+				var buf bytes.Buffer
+				n, err := store.Read(path, &buf)
+
+				Convey("The error should be nil", func() {
+					So(err, ShouldBeNil)
+				})
+
+				Convey("The content should be match", func() {
+					// The content should match the secondTarget
+					So(buf, ShouldNotBeNil)
+					So(n, ShouldEqual, secondSize)
+					So(md5.Sum(buf.Bytes()), ShouldResemble, md5.Sum(secondContent))
+				})
+			})
+		})
+	})
+}
+
+func TestLinkerWithLink(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", func() {
+		l, ok := store.(types.Linker)
+		So(ok, ShouldBeTrue)
+
+		d := store.(types.Direr)
+
+		Convey("When CreateLink to an existing dir", func() {
+			size := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
+			r := io.LimitReader(randbytes.NewRand(), size)
+			target := uuid.New().String()
+
+			_, err := store.Write(target, r, size)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(target)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			path := uuid.New().String()
+			_, err = d.CreateDir(path)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(path, pairs.WithObjectMode(types.ModeDir))
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			o, err := l.CreateLink(path, target)
+
+			Convey("The error should be ErrObjectModeInvalid", func() {
+				So(errors.Is(err, services.ErrObjectModeInvalid), ShouldBeTrue)
+			})
+
+			Convey("The object should be nil", func() {
+				So(o, ShouldBeNil)
+			})
+		})
+	})
+}
+
+func TestLinkerWithVirtualLink(t *testing.T, store types.Storager) {
```

Comment:
This function will be called by services that support `overwrite`, like `oss`. But actually `symlink` is native support for `oss`. So maybe the function name `TestLinkerWithVirtualLink` is not suitable?

Reply:
How about  using ` TestLinkerWithoutOverwrite` and ` TestLinkerWithOverWrite`  as function names?